src/foam/parse/DateParser.java

```diff
@@ -0,0 +1,229 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.parse;
+
+import java.util.Calendar;
+import java.util.Date;
+
+import foam.lib.json.IntParser;
+import foam.lib.json.LongParser;
+import foam.lib.json.Whitespace;
+import foam.lib.parse.Alt;
+import foam.lib.parse.Literal;
+import foam.lib.parse.Optional;
+import foam.lib.parse.PStream;
+import foam.lib.parse.ParserContext;
+import foam.lib.parse.ProxyParser;
+import foam.lib.parse.Seq;
+
+//YYYY
+//YYYY-MM               YYYY/MM
+//YYYY-MM-DD            YYYY/MM/DD
+//YYYY-MM-DDTHH         YYYY/MM/DDTHH
+//YYYY-MM-DDTHH:MM      YYYY/MM/DDTHH:MM
+//YYYY-MM-DD..YYYY-MM-DDYYYY-MM-DD..YYYY-MM-DD
+//today
+//today-7
+public class DateParser
+  extends ProxyParser
+{
+  public DateParser() {
+    super(
+        new Alt(
+
+          //YYYY-MM-DD..YYYY-MM-DD
+          new Seq(
```

Comment:
You have set up a bit Alt() of all the different date formats.  Let's move each potential date format such as "YYYY-MM-DD..YYYY-MM-DD" or "today-7" to their own Parser objects.  Then this class can just become

new Alt(
  new YYYMMDDParser(),
  new YYYYMMDDRangeParser(),
  ...);

That way the body of this parser doesn't have to do a bunch of conditionals to detect what the format of the query was.

You should try to come up with reasonable names for all the sub cases.